src/workerd/server/workerd-api.c++

```diff
@@ -1073,14 +1075,32 @@ kj::Own<jsg::modules::ModuleRegistry> WorkerdApi::initializeBundleModuleRegistry
     const auto diskCacheSpecifier = "pyodide-internal:disk_cache"_url;
     const auto limiterSpecifier = "pyodide-internal:limiter"_url;
 
-    auto pythonRelease = KJ_ASSERT_NONNULL(getPythonSnapshotRelease(featureFlags));
     // Inject metadata that the entrypoint module will read.
+    auto pythonRelease = KJ_ASSERT_NONNULL(getPythonSnapshotRelease(featureFlags));
+    kj::OneOf<kj::Own<PyodideMetadataReader::State>, jsg::Ref<PyodideMetadataReader>>
+        metadataReader = makePyodideMetadataReader(conf, pythonConfig, pythonRelease);
     pyodideBundleBuilder.addSynthetic(metadataSpecifier,
         jsg::modules::Module::newJsgObjectModuleHandler<api::pyodide::PyodideMetadataReader,
             JsgWorkerdIsolate_TypeWrapper>(
-            [metadataReader = makePyodideMetadataReader(conf, pythonConfig, pythonRelease)](
+            [metadataReader = kj::mv(metadataReader)](
                 jsg::Lock& js) mutable -> jsg::Ref<api::pyodide::PyodideMetadataReader> {
-      return metadataReader.addRef();
+      // The metadata reader state is initially created outside of the isolate lock.
+      // Then, once the module is actually evaluated, we'll lazily create the actual
+      // PyodideMetadataReader instance and pass ownership of the state to it.
+      // In practice this is likely only expected to be called once, but with the
+      // new module registry implementation it is at least *possible* for it to
+      // be called multiple times. We don't want to create a new instance each
+      // time so we'll cache the instance and return a ref on each subsequent call.
+      KJ_SWITCH_ONEOF(metadataReader) {
+        KJ_CASE_ONEOF(state, kj::Own<PyodideMetadataReader::State>) {
+          auto ret = js.alloc<PyodideMetadataReader>(kj::mv(state));
+          metadataReader = ret.addRef();
+          return kj::mv(ret);
+        }
+        KJ_CASE_ONEOF(reader, jsg::Ref<PyodideMetadataReader>) {
+          return reader.addRef();
+        }
+      }
```

Comment:
```suggestion
      }
      KJ_UNREACHABLE;
```

Reply:
heh, spotted at the same time. Constantly forgetting that the default local bazel build doesn't treat warnings as errors. Sigh.